README.md

```diff
@@ -1,209 +1,404 @@
+  rpc Get(EchoRequest) returns (EchoResponse) {}
+}
+
+// The EchoRequest protobuf message definition.
+message EchoRequest {
+  // The text of a message to be echoed.
+  string text = 1;
+}
+
+// The EchoResponse protobuf message definition.
+message EchoResponse {
+  // The text of an echo response.
+  string text = 1;
+}
+```
+
+#### Generating the Service and Client Code
+
+Once we have defined our service we can generate the necessary code to implement
+our service and client.
+
+Models (such as `EchoRequest` and `EchoResponse`) are generated using the
+`protoc` plugin provided by [SwiftProtobuf][swift-protobuf]. Assuming the
+above definition for the Echo service and models were saved as `echo.proto` and
+the `protoc-gen-swift` plugin is available in your `PATH` then the following
+will generate the models and write them to `echo.pb.swift` in the current
+directory:
+
+```sh
+protoc echo.proto --swift_out=.
+```
+
+Swift gRPC, provides a plugin (`protoc-gen-swiftgrpc`) to generate the client
+and server for the `Echo` service defined above. It can be invoked to produce
+`echo.grpc.swift` as such:
+
+```sh
+protoc echo.proto --swiftgrpc_out=.
+```
+
+By default both the client and service code is generated (see [Plugin
+Options](#plugin-options) for more details).
+
+#### Implementing the Service Code
+
+The generated service code includes a protocol called `Echo_EchoProvider` which
+defines the set of RPCs that the service offers. The service can be implemented
+by creating a class which conforms to the protocol.
+
+```swift
+import GRPC
+import NIO
+
+class EchoProvider: Echo_EchoProvider {
+  func get(request: Echo_EchoRequest, context: StatusOnlyCallContext) -> EventLoopFuture<Echo_EchoResponse> {
+    let response = Echo_EchoResponse.with {
+      $0.text = "Swift echo get: \(request.text)"
+    }
+    return context.eventLoop.makeSucceededFuture(response)
+  }
+}
+```
+
+#### Using the Service Provider and Generated Client
+
+Now that we have implemented our service code and generated a client, we can put
+it all together.
+
+First we need the appropriate imports:
+
+```swift
+import GRPC
+import NIO
+import Foundation
+```
+
+Since this is just a locally hosted example we'll run the server on localhost
+port 8080.
+
+```
+let host = "localhost"
+let port = 8080
+```
+
+First we need to start the server and provide the Echo service using the
+`EchoProvider` we implemented above. We create a [multhithreaded event loop
+group][nio-ref-elg] which will spawn a single thread to run events on the
+server. Note that you can also use `System.coreCount` to get the number of
+logical cores on your system.
+
+We also `wait` for the server to start before setting up a client.
+
+```swift
+let server = try Server.start(
+  hostname: host,
+  port: port,
+  eventLoopGroup: MultiThreadedEventLoopGroup(numberOfThreads: 1),
+  serviceProviders: [EchoProvider()]
+).wait()
+```
+
+Once the server has started we can create a client by first defining its
```

Comment:
I would make it explicit that the client is usually going to be a different binary.